### PR TITLE
Prioritize Previously-Used GPUs During Auto-Selection

### LINKED ISSUE #10456: When auto-selecting GPUs, prioritize previously used GPUs
## Proposed refactoring or deprecation

When auto-selecting GPUs, Lightning iterates over available CUDA devices in sequence as follows:
https://github.com/PyTorchLightning/pytorch-lightning/blob/4755ae60b71a857a17f39dba1189c2c3245660f5/pytorch_lightning/tuner/auto_gpu_select.py#L48-L58

We should first check GPUs that have been previously used, and then check the unused ones.

### Motivation

Say we have processes A, B and C, and two GPUs configured in exclusive process mode. Consider the following situation:
* Process A is using GPU 0
* Process B, with Lightning, runs a `Trainer` that auto-selects GPU 1 (since A is using GPU 0).
* Process A exits, clearing up GPU 0.
* Process B finished using the `Trainer`, and thus 'frees' up GPU 1. However, B doesn't exit, and so there is still some small amount of memory reserved by PyTorch on GPU 1.
* Process B creates a new `Trainer`, that invokes the auto-selection again. It sees that GPU 0 is available, so it reserves it.
* Process C wants to get a GPU. However, since B is fully using GPU 0, and it's reserved GPU 1 (even though it's not using it), it doesn't get any GPU.

### Pitch

The GPU selection routine should be changed to:
```python
previously_used_gpus = []
unused_gpus = []
for i in range(torch.cuda.device_count()):
    if torch.cuda.memory_reserved(f"cuda:{i}") > 0:
        previously_used_gpus.append(i)
    else:
        unused_gpus.append(i)

# Prioritize previously used GPUs
for i in previously_used_gpus + unused_gpus:
    if i in exclude_gpus:
        continue
    # Try to allocate on device:
    device = torch.device(f"cuda:{i}")
    try:
        torch.ones(1).to(device)
    except RuntimeError:
        continue
    return i
raise RuntimeError("No GPUs available.")
```

I've made the changes here: https://github.com/rharish101/pytorch-lightning/tree/feature/10456-better_gpu_select

cc @justusschock @awaelchli @akihironitta

### PR DESCRIPTION
**Pull request by rharish101**
## What does this PR do?

Fixes #10456

This PR makes the GPU auto-selection function prioritize previously used GPUs. With this change, when we are running auto-selection for the first time, nothing changes. But when we're running it at a later time, this code will try to choose GPUs that were previously used by the current PyTorch process.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/tuner/auto_gpu_select.py [hunk 1]
```diff
@@ -45,7 +45,16 @@ def pick_single_gpu(exclude_gpus: List[int]) -> int:
...
+        if torch.cuda.memory_reserved(f"cuda:{i}") > 0:
+            previously_used_gpus.append(i)
+        else:
+            unused_gpus.append(i)
+
+    # Prioritize previously used GPUs
+    for i in previously_used_gpus + unused_gpus:
         if i in exclude_gpus:
             continue
         # Try to allocate on device:
```
Comment (rohitgr7) [inline pytorch_lightning/tuner/auto_gpu_select.py:60]:
```suggestion
    for i in range(torch.cuda.device_count()):
        if i in exclude_gpus:
            continue

        if torch.cuda.memory_reserved(f"cuda:{i}") > 0:
            previously_used_gpus.append(i)
        else:
            unused_gpus.append(i)

    # Prioritize previously used GPUs
    for i in previously_used_gpus + unused_gpus:
        # Try to allocate on device:
```
i'd do this :)

**Review by rohitgr7 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by ethanwharris — APPROVED** [review]
